crates/cranelift/src/func_environ.rs

```diff
@@ -698,18 +708,14 @@ impl<'module_environment> FuncEnvironment<'module_environment> {
         } else if pointer_type.bits() > desired_type.bits() {
             pos.ins().ireduce(desired_type, val)
         } else {
-            // Note that we `sextend` instead of the probably expected
-            // `uextend`. This function is only used within the contexts of
-            // `memory.size` and `memory.grow` where we're working with units of
-            // pages instead of actual bytes, so we know that the upper bit is
-            // always cleared for "valid values". The one case we care about
-            // sextend would be when the return value of `memory.grow` is `-1`,
-            // in which case we want to copy the sign bit.
-            //
-            // This should only come up on 32-bit hosts running wasm64 modules,
-            // which at some point also makes you question various assumptions
-            // made along the way...
-            pos.ins().sextend(desired_type, val)
+            // We have a 64-bit memory on a 32-bit host -- this combo doesn't
+            // really make a whole lot of sense to do from a user perspective
+            // but that is neither here nor there. We want to unsigned extend
+            // unless `val` is `-1`, as described in the doc comment above.
+            let extended = pos.ins().uextend(desired_type, val);
+            let neg_one = pos.ins().iconst(desired_type, -1);
+            let is_failure = pos.ins().icmp_imm(IntCC::Equal, val, -1);
+            pos.ins().select(is_failure, neg_one, extended)
```

Comment:
Yeah with single-byte pages `-1` is now a valid return from a `memory.{grow,size}` in certain cases and so in this scenario we can't blindly sign extend. I can factor this out to another case for only when the memory has a 1-byte page size.

Reply:
I think sign extension still works since you can't actually consume the entire address space for wasm though, right?

Reply:
But you could consume ~half of it and therefore have the sign bit set, which if we sign extend will result in a negative value / wrong unsigned value.

Reply:
The only "memory growth failed" value though is -1, not any negative number?

Reply:
Yes, but you could have a valid memory size that has the sign bit set now, where that was impossible before because the maximum number of wasm pages that could fit in the address space wasn't large enough. That is, `(u32::MAX / 2) + 1` is a valid memory size with single-byte page sizes, and it has its high bit set, so if we sign-extend that value then we end up giving an incorrect memory size to the Wasm.

Reply:
Right! I'm apparently forgetting how negative numbers work...